flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/DataSourceNode.java

```diff
@@ -187,6 +194,10 @@ else if (len >= 0) {
 					this.estimatedNumRecords = card;
 				}
 			}
+			// close the instantiated resources
+			if(format != null && format instanceof RichInputFormat) {
```

Comment:
IF should be closed after the stats have been retrieved (`format.getStatistics(cachedStatistics);`).